source/base/function_lib.cc

```diff
@@ -2483,6 +2483,130 @@ namespace Functions
     return interpolate (data_values, ix, p_unit);
   }
 
+  /* ---------------------- Polynomial ----------------------- */
+
+
+
+  template <int dim>
+  Polynomial<dim>::
+  Polynomial(const Table<2,double> &exponents,
+             const Vector<double>   &coefficients,
+             const unsigned int   n_components)
+    :
+    Function<dim> (n_components),
+    exponents (exponents),
+    coefficients(coefficients)
+  {
+    Assert(exponents.n_rows() == coefficients.size(),
+           ExcDimensionMismatch(exponents.n_rows(), coefficients.size()));
+  }
+
+
+
+  template <int dim>
+  double
+  Polynomial<dim>::value (const Point<dim> &p,
+                          const unsigned int component) const
+  {
+    (void)component;
+    Assert (component<this->n_components,
+            ExcIndexRange(component, 0, this->n_components)) ;
+
+    double prod;
+    double sum = 0;
+    for (unsigned int monom = 0; monom < exponents.n_rows(); ++monom)
+      {
+        prod = 1;
+        for (unsigned int s=0; s< exponents.n_cols(); ++s)
+          {
+            if (p[s] < 0)
+              Assert(std::floor(exponents[monom][s]) == exponents[monom][s],
+                     ExcMessage("Exponentiation of a negative base number with "
+                                "a real exponent can't be performed."));
+            prod *= std::pow(p[s], exponents[monom][s]);
+          }
+        sum = sum + coefficients[monom]*prod;
+      }
+    return sum;
+  }
+
+
+
+  template <int dim>
+  void
+  Polynomial<dim>::vector_value (const Point<dim> &p,
+                                 Vector<double> &values) const
+  {
+    Assert (values.size() == this->n_components,
+            ExcDimensionMismatch (values.size(), this->n_components));
+
+    for (unsigned int i=0; i<values.size(); ++i)
+      values(i) = Polynomial<dim>::value(p,i);
+  }
+
+
+
+  template <int dim>
+  void
+  Polynomial<dim>::value_list (const std::vector<Point<dim> > &points,
+                               std::vector<double>    &values,
+                               const unsigned int     component) const
+  {
+    Assert (values.size() == points.size(),
+            ExcDimensionMismatch(values.size(), points.size()));
+
+    for (unsigned int i=0; i<points.size(); ++i)
+      values[i] = Polynomial<dim>::value (points[i], component);
+  }
+
+
+
+  template <int dim>
+  Tensor<1,dim>
+  Polynomial<dim>::gradient (const Point<dim> &p,
+                             const unsigned int component) const
+  {
+    (void)component;
+    Assert (component==0, ExcIndexRange(component,0,1));
+
+    Tensor<1,dim> r;
+
+    for (unsigned int d=0; d<dim; ++d)
+      {
+        double sum = 0;
+
+        // This if is placed to allow the use of tables of exponents with n_cols < dim.
+        // The exponents in the remaining dimensions are assumed to be equal to zero.
+        if (d < exponents.n_cols())
```

Comment:
why allow this here but not in `value()`? I think it's not a burden to the user to always pass full tables with dim columns, so I would remove this functionality.